File: src/domain/rules.ts
import type { DamageTrack, PipState } from './types';

export const BOUNDS = {
  hungerMax: 5,
  humanityMax: 10,
  stainsMax: 10,
  damageMinMax: 1,
  damageMaxMax: 20,
} as const;

export function clamp(n: number, min: number, max: number): number {
  if (Number.isNaN(n)) return min;
  return Math.max(min, Math.min(max, Math.trunc(n)));
}

export function clampTrack(track: DamageTrack): DamageTrack {
  const max = clamp(track.max, BOUNDS.damageMinMax, BOUNDS.damageMaxMax);
  const aggravated = clamp(track.aggravated, 0, max);
  const superficial = clamp(track.superficial, 0, max - aggravated);
  return { max, superficial, aggravated };
}

/** Total filled boxes (superficial + aggravated). */
export function totalDamage(track: DamageTrack): number {
  return track.superficial + track.aggravated;
}

/** A damage track is "impaired" when all boxes are filled. */
export function isImpaired(track: DamageTrack): boolean {
  return totalDamage(track) >= track.max;
}

/** A health track is at "torpor"/incapacitated when every box is aggravated. */
export function isTorpor(track: DamageTrack): boolean {
  return track.aggravated >= track.max && track.max > 0;
}

/** Hunger severity tier, for styling. */
export function hungerSeverity(hunger: number): 'normal' | 'high' | 'critical' {
  if (hunger >= BOUNDS.hungerMax) return 'critical';
  if (hunger >= 4) return 'high';
  return 'normal';
}

/** Humanity severity tier, for styling. */
export function humanitySeverity(humanity: number): 'normal' | 'low' | 'critical' {
  if (humanity <= 2) return 'critical';
  if (humanity <= 4) return 'low';
  return 'normal';
}

/** Stains + Humanity > 10 signals degeneration risk. */
export function atDegenerationRisk(humanity: number, stains: number): boolean {
  return humanity + stains > BOUNDS.humanityMax;
}

/**
 * Compute pip states for a damage track, rendered left-to-right.
 * Convention: aggravated fill from the left end of the "damaged" region,
 * superficial follow, then empty boxes. This matches V5 sheets.
 */
export function pipStatesFor(track: DamageTrack): PipState[] {
  const { max, superficial, aggravated } = clampTrack(track);
  const states: PipState[] = [];
  for (let i = 0; i < max; i++) {
    if (i < aggravated) states.push('aggravated');
    else if (i < aggravated + superficial) states.push('superficial');
    else states.push('empty');
  }
  return states;
}

/**
 * Apply a tap at index `i` using the cycle: empty → superficial → aggravated → empty.
 * We convert the tap into a +/- adjustment of the superficial/aggravated counts so
 * the resulting track remains a valid (aggregated, left-aligned) track.
 */
export function cycleAt(track: DamageTrack, index: number): DamageTrack {
  const states = pipStatesFor(track);
  const current = states[index];
  if (current === undefined) return track;

  let { superficial, aggravated } = track;
  if (current === 'empty') {
    // add one superficial
    if (superficial + aggravated < track.max) superficial += 1;
  } else if (current === 'superficial') {
    // convert one superficial into aggravated
    if (superficial > 0) {
      superficial -= 1;
      aggravated += 1;
    }
  } else {
    // aggravated → empty (remove that aggravated box)
    if (aggravated > 0) aggravated -= 1;
  }
  return clampTrack({ max: track.max, superficial, aggravated });
}

export function adjustDamage(
  track: DamageTrack,
  kind: 'superficial' | 'aggravated',
  delta: number,
): DamageTrack {
  const next = { ...track };
  if (kind === 'superficial') next.superficial += delta;
  else next.aggravated += delta;
  return clampTrack(next);
}

export function clearDamage(track: DamageTrack): DamageTrack {
  return { ...track, superficial: 0, aggravated: 0 };
}

File: android/app/src/main/kotlin/io/crimsonledger/domain/Rules.kt
package io.crimsonledger.domain

/**
 * Pure functions operating on the domain model. No Android / Compose types here.
 */

object Bounds {
    const val HUNGER_MAX = 5
    const val HUMANITY_MAX = 10
    const val TRACK_MIN = 1
    const val TRACK_MAX = 20
}

fun clamp(value: Int, min: Int, max: Int): Int =
    if (value < min) min else if (value > max) max else value

/**
 * Aggravated displaces superficial at the cap. Mirrors the web rule.
 */
fun clampTrack(track: DamageTrack): DamageTrack {
    val max = clamp(track.max, Bounds.TRACK_MIN, Bounds.TRACK_MAX)
    val aggravated = clamp(track.aggravated, 0, max)
    val superficial = clamp(track.superficial, 0, max - aggravated)
    return DamageTrack(max = max, superficial = superficial, aggravated = aggravated)
}

/**
 * Produces a left-to-right pip list. Aggravated first, then superficial, then empty —
 * matches web ordering so the leftmost filled pip is always the "worst".
 */
fun pipStatesFor(track: DamageTrack): List<PipState> {
    val t = clampTrack(track)
    return buildList(t.max) {
        repeat(t.aggravated) { add(PipState.AGGRAVATED) }
        repeat(t.superficial) { add(PipState.SUPERFICIAL) }
        repeat(t.max - t.aggravated - t.superficial) { add(PipState.EMPTY) }
    }
}

/**
 * Tap-to-cycle semantics:
 *  - empty     → superficial (added)
 *  - superficial → aggravated (one sup becomes agg)
 *  - aggravated  → empty (one agg removed)
 * Operates by position in [pipStatesFor].
 */
fun cycleAt(track: DamageTrack, index: Int): DamageTrack {
    val states = pipStatesFor(track)
    if (index !in states.indices) return track
    val state = states[index]
    val next = when (state) {
        PipState.EMPTY -> track.copy(superficial = track.superficial + 1)
        PipState.SUPERFICIAL -> track.copy(
            superficial = track.superficial - 1,
            aggravated = track.aggravated + 1,
        )
        PipState.AGGRAVATED -> track.copy(aggravated = track.aggravated - 1)
    }
    return clampTrack(next)
}

enum class DamageKind { SUPERFICIAL, AGGRAVATED }

fun adjustDamage(track: DamageTrack, kind: DamageKind, delta: Int): DamageTrack {
    val next = when (kind) {
        DamageKind.SUPERFICIAL -> track.copy(superficial = track.superficial + delta)
        DamageKind.AGGRAVATED -> track.copy(aggravated = track.aggravated + delta)
    }
    return clampTrack(next)
}

fun clearDamage(track: DamageTrack): DamageTrack =
    track.copy(superficial = 0, aggravated = 0)

fun isImpaired(track: DamageTrack): Boolean {
    val t = clampTrack(track)
    return t.superficial + t.aggravated >= t.max
}

fun isTorpor(track: DamageTrack): Boolean {
    val t = clampTrack(track)
    return t.aggravated >= t.max
}

fun hungerSeverity(thirst: Int): Severity = when {
    thirst >= Bounds.HUNGER_MAX -> Severity.DANGER
    thirst >= 4 -> Severity.WARN
    else -> Severity.NONE
}

fun humanitySeverity(morality: Int): Severity = when {
    morality <= 2 -> Severity.DANGER
    morality <= 4 -> Severity.WARN
    else -> Severity.NONE
}

fun atDegenerationRisk(morality: Int, marks: Int): Boolean =
    (morality + marks) > Bounds.HUMANITY_MAX
